yolo_detection_system/modules/BackgroundDirectionDetector.cpp

```diff
@@ -0,0 +1,249 @@
+
+  // 推論の結果をfloatのベクター配列に変換
+  vector<vector<float>> results;
+  results.reserve(outputTensors.size());
+
+  for(auto& tensor : outputTensors) {
+    float* floatArray = tensor.GetTensorMutableData<float>();
+    size_t totalSize = tensor.GetTensorTypeAndShapeInfo().GetElementCount();
+    results.emplace_back(floatArray, floatArray + totalSize);
+  }
+
+  return results;
+}
+
+void BackgroundDirectionDetector::postprocess(const vector<vector<float>>& outputs,
+                                              const Mat& frame, float scale, int padX, int padY)
+{
+  vector<int> classIds;       // 最も高いスコアを持つクラスIDを格納するリスト
+  vector<float> confidences;  // 信頼度を格納するリスト
+  vector<Rect> boxes;         // バウンディングボックスを格納するリスト
+
+  if(outputs.empty()) {
+    outputError();
+    return;
+  }
+
+  const vector<float>& data = outputs[0];
+
+  // YOLO11 出力 shape: [1, 8, N]
+  int attributes = 8;
+  int numClasses = 8 - 4;  // 座標4つ以外はクラススコア
+  int numBoxes = data.size() / attributes;
+
+  // 検出候補の数だけループを回す
+  for(int i = 0; i < numBoxes; i++) {
+    // クラススコアの最大値とクラスIDを取得
+    float maxScore = -1.0f;
+    int bestClass = -1;
+    for(int j = 0; j < numClasses; j++) {
+      float score = data[i + numBoxes * (4 + j)];
+      if(score > maxScore) {
+        maxScore = score;
+        bestClass = j;
+      }
+    }
+
+    // 最大クラススコアが閾値を超えているかチェック
+    if(maxScore < CONFIDENCE_THRESHOLD) continue;
+
+    float cx = data[i];
+    float cy = data[i + numBoxes];
+    float w = data[i + numBoxes * 2];
+    float h = data[i + numBoxes * 3];
+
+    // バウンディングボックスの中心座標とサイズを表す値
+    int centerX = static_cast<int>((cx - padX) / scale);
+    int centerY = static_cast<int>((cy - padY) / scale);
+    int width = static_cast<int>(w / scale);
+    int height = static_cast<int>(h / scale);
+
+    int left = centerX - width / 2;
+    int top = centerY - height / 2;
+
+    // 画面内に収まるように処理
+    left = max(0, min(left, frame.cols - 1));
+    top = max(0, min(top, frame.rows - 1));
+    width = min(width, frame.cols - left);
+    height = min(height, frame.rows - top);
+
+    // リストに追加
+    boxes.emplace_back(left, top, width, height);
+    confidences.push_back(maxScore);
+    classIds.push_back(bestClass);
+  }
+
+  // 信頼度を超えるバウンディングボックスがなかった時の処理
+  if(boxes.empty()) {
+    outputError();
+    return;
+  }
+
+  // Non-Maximum Suppression (NMS) を実行し、重複する検出ボックスを削減する
+  vector<int> indices;
+  NMSBoxes(boxes, confidences, CONFIDENCE_THRESHOLD, NMS_THRESHOLD, indices);
+
+  // NMS の結果、検出がなければ処理を終了
+  if(indices.empty()) {
+    outputError();
+    return;
+  }
+
+  // NMS 後の最良の検出ボックスのインデックスを取得
+  int bestIdx = indices[0];
+  string direction;
+
+  // 最良検出ボックスのクラスIDに応じて方向を設定
+  if(classIds[bestIdx] == 0) {
+    direction = "FRONT";
+  } else if(classIds[bestIdx] == 1) {
+    direction = "BACK";
+  } else if(classIds[bestIdx] == 2) {
+    direction = "RIGHT";
+  } else if(classIds[bestIdx] == 3) {
+    direction = "LEFT";
+  }
```

Comment:
ここ、向きの型定義を再利用できるとこにおいて、それ使った方が安全だと思う。
```suggestion
  if(classIds[bestIdx] == MiniFigDirection::FRONT) {
    direction = "FRONT";
  }
```
みたいに

Reply:
yolo_detection_system 側では、MiniFigDirectionを定義してないから使えないんだけど、それをどこかに定義するってこと？

Reply:
すでに定義してる型だけを他のファイルに切り出して、インクルードに含めるとかかな。

Reply:
これやろうと試みたんだけど、YOLOのモデルの関係でできない感じでした。

理由としては、YOLO の判定結果のクラス分けが実際の撮影ポイントとかみ合っていないから、撮影ポイントに合わせて定義している向きの型を、判定結果との照らし合わせでつかっちゃうと実際の判定結果と違う向きになってしまうことなんよね

で、これを解決するためには、モデル側のクラスを撮影ポイントに合わせる形に書き換える（たぶん再学習する）必要があるから、今回はちょっとこの形で勘弁願いたいです

こうすればいいんでねって案があったら募集中



**以下詳細（読まなくてもいいやつ）**

YOLO のモデルのクラスって以下になっているんだよね
- class0: FRONT
- class1: BACK
- class2: RIGHT
- class3: LEFT

対して、実際の撮影ポイントと判定結果のかみ合わせが以下
- Point0 → FRONT
- Point1 → RIGHT
- Point2 → BACK
- Point3 → LEFT

で、向き判定結果格納用の列挙体の定義は、実際の撮影ポイントに合わせて、
`enum class MiniFigDirection {FRONT, RIGHT, BACK, LEFT};`
ってしているわけよ

つまり、
```
if(classIds[bestIdx] == MiniFigDirection::FRONT) {
    direction = "FRONT";
  }else if (classIds[bestIdx] == MiniFigDirection::BACK) {
    direction = "BACK";
  }
```
ってしちゃうと`classIds[bestIdx]` が２の時にBACKを結果に入れてしまうことになってしまう（実際のclass 2 は RIGHT）ことになって齟齬が起きてしまうと言った感じです。